### PR TITLE
Add alternative implementation for Mojang's FileIO.

### PR DESCRIPTION
**Pull request by cybertiger**
This probably needs more extensive testing before pulling.

New IO implementation is entirely optional.
Better locking.
Better guarantees about closing region files on world unload.
Better detection and handling of corrupt region files.
Region files can shrink as well as grow.
Better free space map implementation for region files.
Fairly minimal diff.
Configurable thread count for async chunk saving.
Configurable limit to the async chunk saving queue size.

Future: possibility of a plugin API to modify world persistence.
Future: possibility of per world disk IO stats.

### CODE REVIEW

**Comment by zachbr** [comment]
I will try and start playing with this this week. This is obviously something that is going to need extensive testing beyond what any one person can do. Hopefully we can find a medium-large server to take some backups and start experimenting with this.

In the meantime, if you do discover any fixes or we need anything changed, please push them as a separate commits, don't squash. Thanks.

**Comment by NachtRaben** [comment]
AliACraft is a medium-large network that would love to do some testing with this as needed.

**Comment by TitanicFreak** [comment]
We'd happily do some testing over at MineSuperior with this if it's still needed.

**Comment by cybertiger** [comment]
Added fairly basic disk io stats.
iostat command output

File: Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch [hunk 1]
```diff
@@ -0,0 +1,739 @@
...
++        setPermission("bukkit.command.iostat");
++    }
++
++    @Override
++    public boolean execute(CommandSender sender, String commandLabel, String[] args) {
++        if ( !testPermission(sender) ) {
++            return true;
++        }
++
++        if ( !PaperConfig.usePaperIO ) {
```
Comment (kashike) [inline Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch]: No need for weird formatting.

File: Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch [hunk 2]
```diff
@@ -0,0 +1,739 @@
...
++    public IostatCommand(String name) {
++        super(name);
++        this.description = "Get statistics on disk utilisation per world.";
++        this.usageMessage = "/iostat";
++        setPermission("bukkit.command.iostat");
++    }
++
++    @Override
++    public boolean execute(CommandSender sender, String commandLabel, String[] args) {
++        if ( !testPermission(sender) ) {
```
Comment (kashike) [inline Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch]: No need for weird formatting.

File: Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch [hunk 3]
```diff
@@ -0,0 +1,739 @@
...
++            return true;
++        }
++
++        if ( !PaperConfig.usePaperIO ) {
++            sender.sendMessage(ChatColor.RED + "Please enable settings.use-paper-io in order to use iostat.");
++            return true;
++        }
++
++        if ( !PaperConfig.paperIOStats ) {
```
Comment (kashike) [inline Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch]: No need for weird formatting.

File: Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch [hunk 4]
```diff
@@ -0,0 +1,739 @@
...
+@@ -3,9 +3,30 @@ package org.bukkit.craftbukkit.chunkio2;
+ /**
+  * Intended to become public API
+  */
+-public class PersistenceCounters {
++public interface PersistenceCounters {
+ 
+-    // TODO
+-    // Counters pertaining to world saving, so we can get per world stats on chunk io
++    public void bytesRead(long bytes);
```
Comment (kashike) [inline Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch]: `public` is not needed.

File: Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch [hunk 5]
```diff
@@ -0,0 +1,739 @@
...
+new file mode 100644
+index 0000000..40774dd
+--- /dev/null
++++ b/src/main/java/org/bukkit/craftbukkit/chunkio2/CountingOutputStream.java
+@@ -0,0 +1,49 @@
++/*
++ * To change this license header, choose License Headers in Project Properties.
++ * To change this template file, choose Tools | Templates
++ * and open the template in the editor.
++ */
```
Comment (kashike) [inline Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch]: ????????

File: Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch [hunk 6]
```diff
@@ -0,0 +1,739 @@
...
++package org.bukkit.craftbukkit.chunkio2;
++
++/**
++ *
++ * @author antony
++ */
++public class NullPersistenceCounters implements PersistenceCounters {
```
Comment (kashike) [inline Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch]: `final`?

File: Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch (hunk 1 above)
Comment (aikar) [inline Spigot-Server-Patches/0126-Add-an-iostat-command-to-show-disk-usage-stats-per-w.patch]: Also IOStatCommand for consistency.

**Comment by cybertiger** [comment]
I'm probably done with this for now, it works OK and is deployed on a few small servers I run. At this point I doubt there's any performance benefit, the other benefits are still valid.

Since by default the alternate IO impl. is turned off, it's probably safe to pull (though I'd give it a few days, from the latest commit, maybe something will blow up after I deploy it). The only difference is that in once place instead of calling chunkExists() followed by syncChunkLoad() the code just calls syncChunkLoad() (avoids a bunch of duplicated code and lock obtaining since syncChunkLoad() returns null anyway if the chunk does not exist).

I might come back later and fiddle with async chunk saving, but to be honest it'd be huge NMS diff.

**Comment by zachbr** [comment]
I think we can all tell this is awesome, however there are a few issues, which the author has been more than up front about. 

Breaks certain plugins, there is little if any performance improvement, and it's a missed diff waiting to happen on updates.

Thank you anyway Cyber

**Pull request closed without merging**